# **Ejemplos de Test de Hipótesis**
###**Módulo de Nivelación**
###**Técnicas de Inteligencia Computacional I**
###Minor Ciencia de Datos/Minor Ciencia de Datos Avanzado

In [3]:
!pip install scikit_posthocs

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [4]:
import scikit_posthocs as sp
import numpy as np
from scipy import stats
import pandas as pd
pd.options.display.float_format = '{:,.4f}'.format

In [5]:
def check_normality(data):
    test_stat_normality, p_value_normality=stats.shapiro(data)
    print("p value:%.4f" % p_value_normality)
    if p_value_normality <0.05:
        print("Reject null hypothesis >> The data is not normally distributed")
    else:
        print("Fail to reject null hypothesis >> The data is normally distributed")       

In [6]:
def check_variance_homogeneity(group1, group2):
    test_stat_var, p_value_var= stats.levene(group1,group2)
    print("p value:%.4f" % p_value_var)
    if p_value_var <0.05:
        print("Reject null hypothesis >> The variances of the samples are different.")
    else:
        print("Fail to reject null hypothesis >> The variances of the samples are same.")

## P1.
 Un profesor universitario dio conferencias en línea en lugar de clases presenciales debido a Covid-19. Más tarde, subió conferencias grabadas a la nube para los estudiantes que siguieron el curso de forma asincrónica (aquellos que no asistieron a la lección pero luego vieron los registros). Sin embargo, él cree que los estudiantes que asisten a clase en el horario de clase y participan en el proceso tienen más éxito. Por lo tanto, registró las calificaciones promedio de los estudiantes al final del semestre. Los datos están abajo.

synchronous = [94. , 84.9, 82.6, 69.5, 80.1, 79.6, 81.4, 77.8, 81.7, 78.8, 73.2, 87.9, 87.9, 93.5, 82.3, 79.3, 78.3, 71.6, 88.6, 74.6, 74.1, 80.6]      
asynchronous = [77.1, 71.7, 91. , 72.2, 74.8, 85.1, 67.6, 69.9, 75.3, 71.7, 65.7, 72.6, 71.5, 78.2]

**Conducir la prueba de hipótesis para verificar si la creencia del profesor es estadísticamente significativa usando un nivel de significancia de 0.05 para evaluar las hipótesis nula y alternativa. Antes de realizar la prueba de hipótesis, verifique los supuestos relacionados. Comenta los resultados.**

### Supuestos
- Las observaciones en cada muestra son independientes e idénticamente distribuidas (iid).
- Las observaciones en cada muestra se distribuyen normalmente.
- Las observaciones en cada muestra tienen la misma varianza.

In [7]:
sync = np.array([94. , 84.9, 82.6, 69.5, 80.1, 79.6, 81.4, 77.8, 81.7, 78.8, 73.2,
       87.9, 87.9, 93.5, 82.3, 79.3, 78.3, 71.6, 88.6, 74.6, 74.1, 80.6])
asyncr =np.array([77.1, 71.7, 91. , 72.2, 74.8, 85.1, 67.6, 69.9, 75.3, 71.7, 65.7, 72.6, 71.5, 78.2])

In [8]:
media_s = sync.mean()
print(media_s)
media_a = asyncr.mean()
print(media_a)

81.01363636363634
74.60000000000001


- $H_{0}$: los datos se distribuyen normalmente.
- $H_{1}$: los datos no se distribuyen normalmente.

Suponga que alfa = 0.05 Si el valor p es> 0.05, se puede decir que los datos se distribuyen normalmente.

In [9]:
check_normality(sync)
check_normality(asyncr)

p value:0.6556
Fail to reject null hypothesis >> The data is normally distributed
p value:0.0803
Fail to reject null hypothesis >> The data is normally distributed


- $H_{0}$: Las varianzas de las muestras son las mismas.
- $H_{1}$: Las varianzas de las muestras son diferentes.
    
Prueba la hipótesis nula de que las varianzas de la población son iguales (llamada homogeneidad de varianza u homocedasticidad). Si el valor p resultante de la prueba de Levene es menor que algún nivel de significancia (típicamente 0.05), es poco probable que las diferencias obtenidas en las varianzas de la muestra hayan ocurrido en base al muestreo aleatorio de una población con varianzas iguales.

In [10]:
check_variance_homogeneity(sync, asyncr)

p value:0.8149
Fail to reject null hypothesis >> The variances of the samples are same.


$H_{0}$: $\mu_{s}<= \mu_{a}$     
$H_{1}$: $\mu_{s}>  \mu_{a}$

In [11]:
ttest,p_value = stats.ttest_ind(sync,asyncr)
print("p value:%.8f" % p_value)
print("since the hypothesis is one sided >> use p_value/2 >> p_value_one_sided:%.4f" %(p_value/2))
if p_value/2 <0.05:
    print("Reject null hypothesis")
else:
    print("Fail to reject null hypothesis") 

p value:0.00753598
since the hypothesis is one sided >> use p_value/2 >> p_value_one_sided:0.0038
Reject null hypothesis


**A este nivel de significación, existe evidencia suficiente para concluir que la nota media de los alumnos que siguen la asignatura de forma sincrónica es superior a la de los asincrónicos**

--------

## P2.
Un pediatra quiere ver el efecto del consumo de fórmula en la ganancia de peso mensual promedio (en gramos) de los bebés. Por esta razón, recopiló datos de tres grupos diferentes. El primer grupo son los niños amamantados exclusivamente (reciben solo leche materna), el segundo grupo son los niños que se alimentan solo con fórmula y el último grupo son los niños amamantados y con fórmula. Estos datos son los siguientes

only_breast=[794.1, 716.9, 993. , 724.7, 760.9, 908.2, 659.3 , 690.8, 768.7,
       717.3 , 630.7, 729.5, 714.1, 810.3, 583.5, 679.9, 865.1]      
   
only_formula=[ 898.8,  881.2,  940.2,  966.2,  957.5, 1061.7, 1046.2,  980.4,
        895.6,  919.7, 1074.1,  952.5,  796.3,  859.6,  871.1 , 1047.5,
        919.1 , 1160.5,  996.9]     
        
both=[976.4, 656.4, 861.2, 706.8, 718.5, 717.1, 759.8, 894.6, 867.6,
       805.6, 765.4, 800.3, 789.9, 875.3, 740. , 799.4, 790.3, 795.2 ,
       823.6, 818.7, 926.8, 791.7, 948.3]  
       
**De acuerdo con esta información, realice la prueba de hipótesis para verificar si existe una diferencia entre la ganancia mensual promedio de estos tres grupos usando un nivel de significancia de 0.05. Si hay una diferencia significativa, realice un análisis adicional para encontrar la causa de la diferencia. Antes de realizar la prueba de hipótesis, verifique los supuestos relacionados. Comenta los resultados.**

In [12]:
only_breast=np.array([794.1, 716.9, 993. , 724.7, 760.9, 908.2, 659.3 , 690.8, 768.7,
       717.3 , 630.7, 729.5, 714.1, 810.3, 583.5, 679.9, 865.1])

only_formula=np.array([ 898.8,  881.2,  940.2,  966.2,  957.5, 1061.7, 1046.2,  980.4,
        895.6,  919.7, 1074.1,  952.5,  796.3,  859.6,  871.1 , 1047.5,
        919.1 , 1160.5,  996.9])

both=np.array([976.4, 656.4, 861.2, 706.8, 718.5, 717.1, 759.8, 894.6, 867.6,
       805.6, 765.4, 800.3, 789.9, 875.3, 740. , 799.4, 790.3, 795.2 ,
       823.6, 818.7, 926.8, 791.7, 948.3])

In [13]:
media_OB = only_breast.mean()
print(media_OB)
media_OF = only_formula.mean()
print(media_OF)
media_B = both.mean()
print(media_B)

749.8235294117648
959.2157894736845
809.9521739130436


- $ H_{0} $: Los datos se distribuyen normalmente.
- $ H_{1} $: Los datos no se distribuyen normalmente.

In [14]:
check_normality(only_breast)
check_normality(only_formula)
check_normality(both)

p value:0.4694
Fail to reject null hypothesis >> The data is normally distributed
p value:0.8879
Fail to reject null hypothesis >> The data is normally distributed
p value:0.7973
Fail to reject null hypothesis >> The data is normally distributed


- $H_{0}$: Las varianzas de las muestras son las mismas.
- $H_{1}$: Las varianzas de las muestras son diferentes.

In [15]:
stat, pvalue_levene= stats.levene(only_breast,only_formula,both)

print("p value:%.4f" % pvalue_levene)
if pvalue_levene <0.05:
    print("Reject null hypothesis >> The variances of the samples are different.")
else:
    print("Fail to reject null hypothesis >> The variances of the samples are same.")

p value:0.7673
Fail to reject null hypothesis >> The variances of the samples are same.


- $H_{0}$: $\mu_{1}= \mu_{2}= \mu_{3} $ **o** La media de las muestras es la misma.
- $H_{1}$: Al menos uno de ellos es diferente.

In [16]:
F, p_value = stats.f_oneway(only_breast,only_formula,both)
print("p value:%.6f" % p_value)
if p_value <0.05:
    print("Reject null hypothesis")
else:
    print("Fail to reject null hypothesis")

p value:0.000000
Reject null hypothesis


**A este nivel de significación, se puede concluir que al menos uno de los grupos tiene una ganancia de peso promedio mensual diferente.**

### post hoc // pairwise comparisons

In [17]:
# pip install scikit-posthocs
# Pairwise T test for multiple comparisons of independent groups. May be used after a parametric ANOVA to do pairwise comparisons.

import scikit_posthocs as sp
posthoc_df= sp.posthoc_ttest([only_breast,only_formula,both], equal_var=True, p_adjust="bonferroni")

group_names= ["only breast", "only formula","both"]
posthoc_df.columns= group_names
posthoc_df.index= group_names
posthoc_df.style.applymap(lambda x: "background-color:violet" if x<0.05 else "background-color: white")

**A este nivel de significación, se puede concluir que**
- "only breast" es diferente que "only formula"
- "only formula" es diferente con ambos "only breast" u "both"
- "both" es diferente que "only formula"

In [18]:
sp.posthoc_ttest([only_breast,only_formula,both], equal_var=True)

1      2      3
1 1.0000 0.0000 0.0432
2 0.0000 1.0000 0.0000
3 0.0432 0.0000 1.0000

-------
## P3.
Un especialista en recursos humanos que trabaja en una empresa de tecnología está interesado en el exceso de tiempo de trabajo de diferentes equipos. Para investigar si existe una diferencia entre las horas extra del equipo de desarrollo de software y el equipo de prueba, seleccionó aleatoriamente a 17 empleados en cada uno de los dos equipos y registró su tiempo extra de trabajo promedio semanal en términos de una hora. Los datos están abajo.

test_team=[6.2,  7.1,  1.5,  2,3 ,  2,  1.5,  6.1,  2.4,  2.3, 12.4,  1.8,  5.3,  3.1, 9.4,  2.3, 4.1]    
software_team=[2.3,  2.1,  1.4,  2.0, 8.7,  2.2,  3.1,  4.2,  3.6, 2.5,  3.1,  6.2, 12.1,  3.9,  2.2, 1.2 ,3.4]

**De acuerdo con esta información, realice la prueba de hipótesis para comprobar si existe una diferencia entre el tiempo de exceso de trabajo de dos equipos utilizando un nivel de significancia de 0,05. Antes de realizar la prueba de hipótesis, verifique los supuestos relacionados. Comenta los resultados**

In [19]:
test_team=np.array([6.2,  7.1,  1.5,  2,3 ,  2,  1.5,  6.1,  2.4,  2.3, 12.4,  1.8,  5.3,  3.1, 9.4,  2.3, 4.1])
developer_team=np.array([2.3,  2.1,  1.4,  2.0, 8.7,  2.2,  3.1,  4.2,  3.6, 2.5,  3.1,  6.2, 12.1,  3.9,  2.2, 1.2 ,3.4])

- $ H_{0} $: Los datos se distribuyen normalmente.
- $ H_{1} $: Los datos no se distribuyen normalmente.

In [20]:
check_normality(test_team)
check_normality(developer_team)

p value:0.0046
Reject null hypothesis >> The data is not normally distributed
p value:0.0005
Reject null hypothesis >> The data is not normally distributed


- $H_{0}$: Las varianzas de las muestras son las mismas.
- $H_{1}$: Las varianzas de las muestras son diferentes.

In [21]:
check_variance_homogeneity(test_team, developer_team)

p value:0.5410
Fail to reject null hypothesis >> The variances of the samples are same.


- $H_{0}$: $\mu_{1}= \mu_{2}$ **o** $\mu_{1}- \mu_{2} = 0 $ **o** La media de las muestras son idénticos.
- $H_{1}$: $\mu_{1} \neq \mu_{2}$ **o** $\mu_{1}- \mu_{2} \neq 0 $ **o** La media de las muestras son diferentes.

In [22]:
ttest,pvalue = stats.mannwhitneyu(test_team,developer_team, alternative="two-sided")
print("p-value:%.4f" % pvalue)
if pvalue <0.05:
    print("Reject null hypothesis")
else:
    print("Fail to recejt null hypothesis")

p-value:0.8226
Fail to recejt null hypothesis


**A este nivel de significancia, se puede decir que no existe una diferencia estadísticamente significativa entre el tiempo de exceso de trabajo promedio de los dos equipos.**

--------
## P4.

Una empresa de comercio electrónico anuncia regularmente sus campañas en YouTube, Instagram y Facebook. Sin embargo, el nuevo gerente tenía curiosidad sobre si había alguna diferencia entre la cantidad de clientes atraídos por estas plataformas. Por eso, empezó a usar Adjust, una aplicación que te permite saber de dónde vienen tus usuarios. Los números diarios informados por Adjust para cada plataforma son los siguientes.

youtube=[1913, 1879, 1939, 2146, 2040, 2127, 2122, 2156, 2036, 1974, 1956, 2146, 2151, 1943, 2125]
       
instagram = [2305., 2355., 2203., 2231., 2185., 2420., 2386., 2410., 2340., 2349., 2241., 2396., 2244., 2267., 2281.]     
       
facebook = [2133., 2522., 2124., 2551., 2293., 2367., 2460., 2311., 2178., 2113., 2048., 2443., 2265., 2095., 2528.]          

In [23]:
youtube=np.array([1913, 1879, 1939, 2146, 2040, 2127, 2122, 2156, 2036, 1974, 1956,
       2146, 2151, 1943, 2125])
       
instagram =  np.array([2305., 2355., 2203., 2231., 2185., 2420., 2386., 2410., 2340.,
       2349., 2241., 2396., 2244., 2267., 2281.])
       
facebook = np.array([2133., 2522., 2124., 2551., 2293., 2367., 2460., 2311., 2178.,
       2113., 2048., 2443., 2265., 2095., 2528.]) 

- $H_{0}$: los datos se distribuyen normalmente.
- $H_{1}$: los datos no se distribuyen normalmente.

In [24]:
check_normality(youtube)
check_normality(instagram)
check_normality(facebook)

p value:0.0285
Reject null hypothesis >> The data is not normally distributed
p value:0.4156
Fail to reject null hypothesis >> The data is normally distributed
p value:0.1716
Fail to reject null hypothesis >> The data is normally distributed


- $H_{0}$: Las varianzas de las muestras son las mismas.
- $H_{1}$: Las varianzas de las muestras son diferentes.

In [25]:
stat, pvalue_levene= stats.levene(youtube, instagram, facebook)

print("p value:%.4f" % pvalue_levene)
if pvalue_levene <0.05:
    print("Reject null hypothesis >> The variances of the samples are different.")
else:
    print("Fail to reject null hypothesis >> The variances of the samples are same.")

p value:0.0012
Reject null hypothesis >> The variances of the samples are different.


- $H_{0}$: $\mu_{1}= \mu_{2}= \mu_{3} $ **o** La media de las muestras es la misma.
- $H_{1}$: Al menos uno de ellos es diferente.

In [26]:
F, p_value = stats.kruskal(youtube, instagram, facebook)
print("p value:%.6f" % p_value)
if p_value <0.05:
    print("Reject null hypothesis")
else:
    print("Fail to reject null hypothesis")

p value:0.000015
Reject null hypothesis


En este nivel de significación, al menos uno de los números promedio de adquisición de clientes es diferente.

**Nota:** Dado que los datos no son normales, se utiliza una versión no paramétrica de la prueba posthoc.

In [27]:
posthoc_df = sp.posthoc_mannwhitney([youtube,instagram, facebook], p_adjust = 'bonferroni')
group_names= ["youtube", "instagram","facebook"]
posthoc_df.columns= group_names
posthoc_df.index= group_names
posthoc_df.style.applymap(lambda x: "background-color:violet" if x<0.05 else "background-color: white")

**El número promedio de clientes que provienen de YouTube es diferente al de los demás (en realidad, más pequeño que los demás).**

-------------------

## P5.

El Centro de Salud METU diagnosticó a dieciocho estudiantes con colesterol alto en el semestre anterior. El personal de atención médica les informó a estos pacientes sobre los peligros del colesterol alto y les prescribió un programa de dieta. Un mes después, los pacientes acudieron para control y se volvió a examinar su nivel de colesterol. Pruebe si hay una diferencia en los niveles de colesterol de los pacientes.

**De acuerdo con esta información, realice la prueba de hipótesis para verificar si hay una disminución en los niveles de colesterol de los pacientes después de la dieta utilizando un nivel de significancia de 0.05. Antes de realizar la prueba de hipótesis, verifique los supuestos relacionados. Comenta los resultados**

test_results_before_diet=[224, 235, 223, 253, 253, 224, 244, 225, 259, 220, 242, 240, 239, 229, 276, 254, 237, 227]  
test_results_after_diet=[198, 195, 213, 190, 246, 206, 225, 199, 214, 210, 188, 205, 200, 220, 190, 199, 191, 218]

## Supuestos
- La variable dependiente debe ser continua (intervalo/razón)
- Las observaciones son independientes entre sí
- La variable dependiente debe tener una distribución aproximadamente normal

In [28]:
test_results_before_diet=np.array([224, 235, 223, 253, 253, 224, 244, 225, 259, 220, 242, 240, 239, 229, 276, 254, 237, 227])
test_results_after_diet=np.array([198, 195, 213, 190, 246, 206, 225, 199, 214, 210, 188, 205, 200, 220, 190, 199, 191, 218])

- $H_{0}$: los datos se distribuyen normalmente.
- $H_{1}$: los datos no se distribuyen normalmente.

In [29]:
check_normality(test_results_before_diet)
check_normality(test_results_after_diet)

p value:0.1635
Fail to reject null hypothesis >> The data is normally distributed
p value:0.1003
Fail to reject null hypothesis >> The data is normally distributed


- $H_{0}$: $\mu_{d}>= 0 $ **o** La verdadera diferencia media es igual o mayor que cero.
- $H_{1}$: $\mu_{d}< 0 $ **o** La verdadera diferencia media es menor que cero.

In [30]:
test_stat, p_value_paired = stats.ttest_rel(test_results_before_diet,test_results_after_diet)
print("p value:%.6f" % p_value_paired , "one tailed p value:%.6f" %(p_value_paired/2))
if p_value_paired <0.05:
    print("Reject null hypothesis")
else:
    print("Fail to reject null hypothesis")

p value:0.000008 one tailed p value:0.000004
Reject null hypothesis


 **En este nivel de significación, hay evidencia suficiente para concluir que el nivel medio de colesterol de los pacientes ha disminuido después de la dieta.**

-----

## P6.
Un capitalista de riesgo quería invertir en una startup que proporciona compresión de datos sin pérdida de calidad, pero hay dos competidores: PiedPiper y EndFrame. Inicialmente, creía que el rendimiento de EndFrame podría ser mejor, pero aun así quería probarlo antes de la inversión. Luego, entregó los mismos archivos a cada empresa para que los comprimieran y registraran sus puntajes de desempeño. Los datos están abajo.
    
piedpiper=[4.57, 4.55, 5.47, 4.67, 5.41, 5.55, 5.53, 5.63, 3.86, 3.97, 5.44, 3.93, 5.31, 5.17, 4.39, 4.28, 5.25]     
endframe = [4.27, 3.93, 4.01, 4.07, 3.87, 4.  , 4.  , 3.72, 4.16, 4.1 , 3.9 , 3.97, 4.08, 3.96, 3.96, 3.77, 4.09]


**De acuerdo con esta información, realice las pruebas de hipótesis relacionadas utilizando un nivel de significación de 0,05. Antes de realizar la prueba de hipótesis, verifique los supuestos relacionados. Comenta los resultados.**

## Supuestos
- La variable dependiente debe ser continua (intervalo/razón)
- Las observaciones son independientes entre sí
- La variable dependiente debe tener una distribución aproximadamente normal

- $H_{0}$: los datos se distribuyen normalmente.
- $H_{1}$: los datos no se distribuyen normalmente.

Suponga que alfa = 0.05 Si el valor p es> 0.05, se puede decir que los datos tienen una distribución normal.

In [31]:
piedpiper=np.array([4.57, 4.55, 5.47, 4.67, 5.41, 5.55, 5.53, 5.63, 3.86, 3.97, 5.44, 3.93, 5.31, 5.17, 4.39, 4.28, 5.25])
endframe = np.array([4.27, 3.93, 4.01, 4.07, 3.87, 4.  , 4.  , 3.72, 4.16, 4.1 , 3.9 , 3.97, 4.08, 3.96, 3.96, 3.77, 4.09])
check_normality(piedpiper)
check_normality(endframe)

p value:0.0304
Reject null hypothesis >> The data is not normally distributed
p value:0.9587
Fail to reject null hypothesis >> The data is normally distributed


- $H_{0}$: $\mu_{d} >= 0 $ **o** La verdadera diferencia media es igual o mayor que cero.
- $H_{1}$: $\mu_{d} < 0 $ **o** La verdadera diferencia media es menor que cero.

In [32]:
test,pvalue = stats.wilcoxon(endframe,piedpiper) ##alternative default two sided
print("p-value:%.6f" %pvalue, ">> one_tailed_pval:%.6f" %(pvalue/2))

test,one_sided_pvalue = stats.wilcoxon(endframe,piedpiper, alternative="less")
print("one sided pvalue:%.6f" %(one_sided_pvalue))
if pvalue <0.05:
    print("Reject null hypothesis")
else:
    print("Fail to recejt null hypothesis")

p-value:0.000214 >> one_tailed_pval:0.000107
one sided pvalue:0.000107
Reject null hypothesis


**Rechazar $H_{0}$ >> En este nivel de significación, hay evidencia suficiente para concluir que el rendimiento de PiedPaper es mejor que el de EndFrame.**

----

## P7.

Una investigadora tenía curiosidad acerca de si existe una diferencia entre la metodología que desarrolló, C, y los métodos de referencia A y B en términos de rendimiento. Por lo tanto, decidió diseñar diferentes experimentos y registró la precisión alcanzada por cada método. La siguiente tabla muestra la precisión alcanzada en los conjuntos de prueba por cada método. Tenga en cuenta que se utilizaron los mismos conjuntos de prueba y entrenamiento para cada método.

| Experiment |   A  |   B  |   C  |
|:----------:|:----:|:----:|:----:|
|     E1     | 89.8 | 90.0 | 91.5 |
|     E2     | 89.9 | 90.1 | 90.7 |
|     E3     | 88.6 | 88.8 | 90.3 |
|     E4     | 88.7 | 88.9 | 90.4 |
|     E5     | 89.6 | 89.9 | 90.2 |
|     E6     | 89.7 | 90.0 | 90.3 |
|     E7     | 89.2 | 89.0 | 90.2 |
|     E8     | 89.3 | 89.2 | 90.3 | 

**De acuerdo con esta información, realice la prueba de hipótesis para verificar si existe una diferencia entre el rendimiento de los métodos usando un nivel de significancia de 0.05. Si hay una diferencia significativa, realice un análisis adicional para encontrar cuál causó la diferencia. Antes de realizar la prueba de hipótesis, verifique los supuestos relacionados. Comenta los resultados.**

## Supuestos
- Las observaciones en cada muestra son independientes e idénticamente distribuidas (iid).
- Las observaciones en cada muestra se distribuyen normalmente.
- Las observaciones en cada muestra tienen la misma varianza.


    
- $H_{0}$: los datos se distribuyen normalmente.
- $H_{1}$: los datos no se distribuyen normalmente

In [33]:
method_A = np.array([89.8, 89.9, 88.6, 88.7, 89.6, 89.7, 89.2, 89.3])
method_B =   np.array([90.0, 90.1, 88.8, 88.9, 89.9, 90.0, 89.0, 89.2])
method_C = np.array([91.5, 90.7, 90.3, 90.4, 90.2, 90.3, 90.2, 90.3])

check_normality(method_A)
check_normality(method_B)
check_normality(method_C)

p value:0.3076
Fail to reject null hypothesis >> The data is normally distributed
p value:0.0515
Fail to reject null hypothesis >> The data is normally distributed
p value:0.0016
Reject null hypothesis >> The data is not normally distributed


- $H_{0}$: Las varianzas de las muestras son las mismas.
- $H_{1}$: Las varianzas de las muestras son diferentes.

In [34]:
stat, pvalue_levene= stats.levene(method_A, method_B, method_C)

print("p value:%.4f" % pvalue_levene)
if pvalue_levene <0.05:
    print("Reject null hypothesis >> The variances of the samples are different.")
else:
    print("Fail to reject null hypothesis >> The variances of the samples are same.")

p value:0.1953
Fail to reject null hypothesis >> The variances of the samples are same.


In [35]:
test_stat,p_value = stats.friedmanchisquare(method_A,method_B, method_C)
print("p value:%.4f" % p_value)
if p_value <0.05:
    print("Reject null hypothesis")
else:
    print("Fail to reject null hypothesis")
    
print(np.round(np.mean(method_A),2), np.round(np.mean(method_B),2), np.round(np.mean(method_C),2))     

p value:0.0015
Reject null hypothesis
89.35 89.49 90.49


### posthoc

In [36]:
data = np.array([method_A, method_B, method_C]) 
posthoc_df=sp.posthoc_wilcoxon(data, p_adjust="holm")
# posthoc_df = sp.posthoc_nemenyi_friedman(data.T) ## another option for the posthoc test

group_names= ["Method A", "Method B","Method C"]
posthoc_df.columns= group_names
posthoc_df.index= group_names
posthoc_df.style.applymap(lambda x: "background-color:violet" if x<0.05 else "background-color: white")

**El método C superó a los demás y logró mejores puntajes de precisión que los demás.**

------------

## P8.

Un analista de una empresa de inversión financiera siente curiosidad por la relación entre el género y el apetito por el riesgo. Se tomó una muestra aleatoria de 660 clientes de la base de datos. Los clientes de la muestra se clasificaron según su género y su apetito por el riesgo. El resultado se da en la siguiente tabla.

| **Gender/Risk Appetite** | Very Low | Low | Medium | High | Very High | Total |
|:--------------------:|:--------:|:---:|:------:|:----:|:---------:|:-----:|
|        **Female**        |    53    |  23 |   30   |  36  |     88    |  230  |
|         **Male**         |    71    |  48 |   51   |  57  |    203    |  430  |
|         **Total**        |    124   |  71 |   81   |  93  |    291    |  660  |

Pruebe la hipótesis de que el apetito por el riesgo de los clientes de esta empresa es independiente de su género. Use α = 0.01.

- $H_{0}$: el género y el apetito por el riesgo son independientes.
- $H_{1}$: el género y el apetito por el riesgo dependen.

Para esta pregunta se debe usar la prueba chi2. Esta prueba se conoce como la prueba de bondad de ajuste. Implica que si los datos observados están muy cerca de los datos esperados. El supuesto de esta prueba cada Ei ≥ 5 (en al menos el 80% de las celdas) que se cumple.

In [37]:
from scipy.stats import chi2_contingency

obs =np.array([[53, 23, 30, 36, 88],[71, 48, 51, 57, 203]])
chi2, p, dof, ex = chi2_contingency(obs, correction=False)

print("expected frequencies:\n ", np.round(ex,2))
print("degrees of freedom:", dof)
print("test stat :%.4f" % chi2)
print("p value:%.4f" % p)

expected frequencies:
  [[ 43.21  24.74  28.23  32.41 101.41]
 [ 80.79  46.26  52.77  60.59 189.59]]
degrees of freedom: 4
test stat :7.0942
p value:0.1310


In [38]:
from scipy.stats import chi2
## calculate critical stat

alpha = 0.01
df = (5-1)*(2-1)
critical_stat = chi2.ppf((1-alpha), df)
print("critical stat:%.4f" % critical_stat)

critical stat:13.2767


**Dado que el valor p es mayor que α=0.01 (o la estadística calculada=7.14 es menor que la estadística crítica=13.28) >> No se puede rechazar H0. A este nivel de significación, se puede concluir que el género y el apetito por el riesgo son independientes.**